File: Numerical_Methods_for_Physics_Garcia/edition2/Cpp/fft2.cpp
#include "NumMeth.h"

void fft( Matrix& RealA, Matrix& ImagA);

void fft2( Matrix& RealA, Matrix& ImagA) {
// Routine to compute two dimensional Fourier transform 
// using FFT algorithm
// Inputs
//    RealA, ImagA         Real and imaginary parts of data array
// Outputs
//    RealA, ImagA         Real and imaginary parts of transform

  int i, j, N = RealA.nRow();
  Matrix RealT(N), ImagT(N);  // Temporary work vector

  //* Loop over the columns of the matrix
  for( j=1; j<=N; j++ ) {
	//* Copy out a column into a vector
    for( i=1; i<=N; i++ ) {
      RealT(i) = RealA(i,j);  
      ImagT(i) = ImagA(i,j);
    }
	//* Take FFT of the vector
    fft(RealT,ImagT);         
	//* Copy the transformed vector back into the column
    for( i=1; i<=N; i++ ) {
      RealA(i,j) = RealT(i); 
      ImagA(i,j) = ImagT(i);
    }
  }

  //* Loop over the rows of the matrix
  for( i=1; i<=N; i++ ) {
	//* Copy out a row into a vector
    for( j=1; j<=N; j++ ) {
      RealT(j) = RealA(i,j);  
      ImagT(j) = ImagA(i,j);
    }
	//* Take FFT of the vector
    fft(RealT,ImagT);         
	//* Copy the transformed vector back into the row
    for( j=1; j<=N; j++ ) {
      RealA(i,j) = RealT(j);  
      ImagA(i,j) = ImagT(j);
    }
  }
}

File: Numerical_Methods_for_Physics_Garcia/edition2/Fortran/fft2.f
      subroutine fft2( A, N, MAXN )
      integer*4 N, MAXN
      complex*16 A(MAXN,MAXN)
! Routine to compute two dimensional Fourier transform
! using FFT algorithm
! Inputs
!    A         Complex input data array
!    N         Elements transformed are A(1,1) to A(N,N)
!   MAXN       Array dimensioned as A(MAXN,MAXN)
! Outputs
!    A         Complex transform of data

      integer*4 MAXNT
      parameter( MAXNT = 2048 )
      integer*4 i, j
      complex*16 T(MAXNT)  ! Temporary work vector

      !* Loop over the columns of the matrix
      do j=1,N
        !* Copy out a column into a vector
        do i=1,N
          T(i) = A(i,j)
        enddo
        !* Take FFT of the vector
        call fft(T,N)
        !* Copy the transformed vector back into the column
        do i=1,N
          A(i,j) = T(i)
        enddo
      enddo

      !* Loop over the rows of the matrix
      do i=1,N
        !* Copy out a row into a vector
        do j=1,N
          T(j) = A(i,j)
        enddo
        !* Take FFT of the vector
        call fft(T,N)
        !* Copy the transformed vector back into the row
        do j=1,N
          A(i,j) = T(j)
        enddo
      enddo
      return
      end
